### DISCUSSION TITLE
Support (porting) @angular-eslint rules to oxlint

### DISCUSSION THREAD

**Discussion by pumano** [angular-eslint/angular-eslint#2062] [General]
Looks like https://github.com/oxc-project/oxc will be next default linter. That linter lint my 90 libs huge workspace for 1 sec instead of many minutes of eslint. According to support for many rules added https://github.com/oxc-project/oxc/issues/481

Have you plans to port / support @angular-eslint and @angular-eslint/template rules to oxlint?

**Comment by JamesHenry** [comment]
Being able to lint Angular templates requires an AST to be produced from Angular flavored HTML. I think it's very unlikely that oxc will be supporting parsing Angular flavored HTML any time soon, so at most you would only be able to port the rules that run on TS constructs, but not anything from `@angular-eslint/eslint-plugin-template`.

I am very excited by oxc, but angular-eslint is one of the last projects that can be ported because of its need for a custom parser for non-JS(X)/TS(X) concerns.

**Reply by JamesHenry** [reply]
If this is something you are passionate about seeing, maybe you can lead the charge for oxc to support pluggability in its parser, so that a relevant rust implementation could be written for turning an Angular Template to an appropriate AST.

**Comment by meDavidNS** [comment]
You can now add this plugin to oxlint using the `jsPlugins` configuration property in `.oxlintrc.json`
```
	"jsPlugins": [
		"@angular-eslint/eslint-plugin",
		"eslint-plugin-tailwindcss",
		"eslint-plugin-simple-import-sort",
		"@nx/eslint-plugin"
	],
```

**Reply by JelleBruisten** [reply]
@meDavidNS could you share your config? I'm running into this error:

` Error: You have used a rule which requires '@angular-eslint/template-parser' to be used as the 'parser' in your ESLint config.`

When trying to run some rules:
```
{
   "jsPlugins": [
       "@angular-eslint/eslint-plugin",
       "@angular-eslint/eslint-plugin-template",
       "@nx/eslint-plugin"
    ],
    "rules": {
      "@angular-eslint/template/button-has-type": "error"
    }
}
```

**Comment by Shinigami92** [comment]
Good news! I got the allowance to spend some dedicated time into OSS contribution to support `@angular-eslint/template` with `oxlint`.
Feel free to come in contact with me, so I know where I can start.

**Reply by JamesHenry** [reply]
It's great that you are enthusiastic about contributing, but it's important to recognise where things stand. Everything I wrote here in October still stands: https://github.com/angular-eslint/angular-eslint/discussions/2062#discussioncomment-10894742

A custom parser is a hard requirement and oxlint does not support custom parsers:
https://github.com/oxc-project/oxc/issues/481#issuecomment-4129313930

And whenever they do introduce that, whoever creates the Angular custom parser is embarking on an _ongoing_ journey of maintaining AST compatibility with the Angular compiler's own AST - every time the Angular Team change something or add new syntax, the oxlint custom parser would need to be updated accordingly.

**Reply by Shinigami92** [reply]
Very good info 👏
If I need to create a repo on my journey, I will try to find a home not under Shinigami92 personal repo, but an GitHub organization which is community driven!

**Reply by pumano** [reply]
https://github.com/voidzero-dev/oxc-angular-compiler this project can help with AST? @JamesHenry

**Reply by Shinigami92** [reply]
I've been mapping out the 41 `@angular-eslint/template` rules to understand what parser features each one requires. Here's a breakdown that could help prioritize work:

### HTML-only (9 rules)
Only need Element, TextAttribute, basic HTML structure.

| Rule | Fix | Summary |
|------|-----|---------|
| button-has-type | — | Ensures buttons have valid type attribute |
| no-autofocus | Fix | Disallows autofocus except on dialogs |
| no-distracting-elements | Fix | Disallows `<blink>` and `<marquee>` |
| no-nested-tags | — | Prevents nesting of `<p>` or `<a>` tags |
| no-positive-tabindex | Suggest | Disallows positive tabindex values |
| prefer-self-closing-tags | Fix | Self-closing tags for empty elements |
| label-has-associated-control | — | Labels must be associated with form controls |
| i18n | Fix | Enforces i18n markers on elements/attributes |
| prefer-static-string-properties | Fix | Use property assignment for static strings |

### Binding (7 rules)
Need BoundAttribute, BoundEvent, or binding expression support.

| Rule | Fix | Summary |
|------|-----|---------|
| banana-in-box | Fix | Detects `([x])` vs `[(x)]` syntax error |
| no-duplicate-attributes | Suggest | Detects duplicate inputs/outputs |
| no-inline-styles | — | Disallows inline styles and style bindings |
| no-interpolation-in-attributes | Fix | Prefer `[attr]` over `{{ }}` in attributes |
| prefer-ngsrc | Suggest | Use ngSrc instead of src for images |
| prefer-class-binding | — | Prefer `[class]` over `[ngClass]` |
| attributes-order | Fix | Enforces consistent attribute ordering |

### Expression (8 rules)
Need to inspect expression AST (PropertyRead, Call, Binary, etc.).

| Rule | Fix | Summary |
|------|-----|---------|
| eqeqeq | Fix | Requires `===`/`!==` over `==`/`!=` |
| no-any | Suggest | Disallows `$any()` type cast |
| no-non-null-assertion | — | Disallows `!` operator in templates |
| no-negated-async | Suggest | Disallows negation of async pipe results |
| no-call-expression | — | Disallows method calls except in event handlers |
| conditional-complexity | — | Limits complexity of conditional expressions |
| prefer-built-in-pipes | — | Prefer built-in pipes over method calls |
| prefer-template-literal | Fix | Prefer template literals over `+` concatenation |

### Control-flow (7 rules)
Need IfBlock, ForLoopBlock, SwitchBlock, DeferredBlock support.

| Rule | Fix | Summary |
|------|-----|---------|
| no-empty-control-flow | — | Detects empty `@if`, `@for`, `@switch` blocks |
| prefer-control-flow | — | Prefer `@if`/`@for`/`@switch` over `*ngIf`/`*ngFor` |
| use-track-by-function | — | Ensures `*ngFor` uses trackBy |
| prefer-at-else | Fix | Prefer `@else` over consecutive opposite `@if` |
| cyclomatic-complexity | — | Measures template control flow complexity |
| prefer-at-empty | Fix | Prefer `@empty` in `@for` over separate `@if` |
| prefer-contextual-for-variables | Fix | Prefer `$index`/`$count`/`$first` etc. over aliases |

### A11y (8 rules)
Accessibility checks — need Element + attributes + ARIA knowledge.

| Rule | Fix | Summary |
|------|-----|---------|
| alt-text | — | img/area/object must have alt text |
| click-events-have-key-events | — | Click handlers need keyboard event pairs |
| mouse-events-have-key-events | — | Mouseover/out need focus/blur pairs |
| elements-content | — | Headings/anchors/buttons must have text content |
| interactive-supports-focus | — | Interactive elements must be focusable |
| role-has-required-aria | Suggest | ARIA roles must have required properties |
| valid-aria | Suggest | Validates ARIA attribute names and values |

The main blocker remains oxc's lack of custom template parser support, as confirmed by @TheAlexLichter. In the meantime, I've started contributing to [oxc-angular-compiler](https://github.com/voidzero-dev/oxc-angular-compiler) which already has a complete Angular HTML + expression parser in Rust — that could serve as the foundation once oxc opens up template parser integration.

**Reply by markgoho** [reply]
angular will probably support tsx before oxlint supports angular

**Reply by ChristofFritz** [reply]
@markgoho stay positive

**Reply by TheAlexLichter** [reply]
@markgoho Supporting custom components is non-trivial. But once the base is there, Vue, Svelte, Angular etc. will all be similarly achievable from what I know! We are on it 😉

**Reply by liangmiQwQ** [reply]
Wow! That's fantastic! I'm doing the similar work for Vue and looking forward to seeing your progress on Angular as well!

**Reply by zargham-k** [reply]
looking forward to this! @TheAlexLichter how to stay up-to-date with what you are cooking?

**Reply by sscowden** [reply]
Any update on this?

**Reply by Shinigami92** [reply]
> Any update on this?

No. We are waiting on generic language/parser support and settled maintainer burden. So this is delayed to 26Q3 or even after 26Q4.

**Reply by sscowden** [reply]
> > Any update on this?
> 
> No. We are waiting on generic language/parser support and settled maintainer burden. So this is delayed to 26Q3 or even after 26Q4.

Ok cool, appreciate the response!
